src/leap/mail/mail.py

```diff
@@ -0,0 +1,697 @@
+            if doc_id is None:
+                return None
+            return self.adaptor.get_msg_from_mdoc_id(
+                self.messageklass, self.store,
+                doc_id, uid=uid, get_cdocs=get_cdocs)
+
+        d = self.mbox_indexer.get_doc_id_from_uid(self.mbox_uuid, uid)
+        d.addCallback(get_msg_from_mdoc_id)
+        return d
+
+    # TODO deprecate ??? ---
+    def _prime_count(self):
+        def update_count(count):
+            self._count = count
+        d = self.mbox_indexer.count(self.mbox_name)
+        d.addCallback(update_count)
+        return d
+
+    def count(self):
+        """
+        Count the messages in this collection.
+        :return: a Deferred that will fire with the integer for the count.
+        :rtype: Deferred
+        """
+        if not self.is_mailbox_collection():
+            raise NotImplementedError()
+
+        d = self.mbox_indexer.count(self.mbox_uuid)
+        return d
+
+    def count_recent(self):
+        # FIXME HACK
+        return 0
+
+    def count_unseen(self):
+        # FIXME hack
+        return 0
+
+    def get_uid_next(self):
+        """
+        Get the next integer beyond the highest UID count for this mailbox.
+
+        :return: a Deferred that will fire with the integer for the next uid.
+        :rtype: Deferred
+        """
+        return self.mbox_indexer.get_next_uid(self.mbox_uuid)
+
+    def all_uid_iter(self):
+        """
+        Iterator through all the uids for this collection.
+        """
+        return self.mbox_indexer.all_uid_iter(self.mbox_uuid)
+
+    # Manipulate messages
+
+    def add_msg(self, raw_msg, flags=tuple(), tags=tuple(), date=""):
+        """
+        Add a message to this collection.
+        """
+        leap_assert_type(flags, tuple)
+        leap_assert_type(date, str)
+
+        msg = self.adaptor.get_msg_from_string(Message, raw_msg)
+        wrapper = msg.get_wrapper()
+
+        if not self.is_mailbox_collection():
+            raise NotImplementedError()
+
+        else:
+            mbox_id = self.mbox_uuid
+            wrapper.set_mbox_uuid(mbox_id)
+            wrapper.set_flags(flags)
+            wrapper.set_tags(tags)
+            wrapper.set_date(date)
+
+        def insert_mdoc_id(_, wrapper):
+            doc_id = wrapper.mdoc.doc_id
+            return self.mbox_indexer.insert_doc(
+                self.mbox_uuid, doc_id)
+
+        d = wrapper.create(self.store)
+        d.addCallback(insert_mdoc_id, wrapper)
+        d.addErrback(lambda f: f.printTraceback())
+
+        return d
+
+    def copy_msg(self, msg, newmailbox):
+        """
+        Copy the message to another collection. (it only makes sense for
+        mailbox collections)
+        """
+        if not self.is_mailbox_collection():
+            raise NotImplementedError()
+
+        def insert_copied_mdoc_id(wrapper):
+            return self.mbox_indexer.insert_doc(
+                newmailbox_uuid, wrapper.mdoc.doc_id)
```

Comment:
Where is this `newmailbox_uuid` coming from? Might it be a typo and you meant `newmailbox`?